pkg/instrument/extension.c

```diff
@@ -0,0 +1,78 @@
+#include "extension.h"
+
+#include <sqlite3.h>
+#include <stdatomic.h>
+#include <stdint.h>
+#include <string.h>
+
+static volatile sqlite3_metrics_t global_metrics = {0};
+
+
+static sqlite3_error auto_instrument_connection(sqlite3 *connection, const char** pzErrMsg, const struct sqlite3_api_routines* pThunk) {
+    return sqlite3_trace_v2(connection, SQLITE_TRACE_PROFILE|SQLITE_TRACE_CLOSE, sqlite3_collect_metrics, NULL);
+}
+
+sqlite3_error sqlite3_instrument() {
+    return sqlite3_auto_extension((void (*)())(auto_instrument_connection));
+}
+
+void sqlite3_deinstrument() {
+    sqlite3_cancel_auto_extension((void (*)())(auto_instrument_connection));
+}
+
+sqlite3_error sqlite3_metrics(sqlite3_metrics_t* metrics, int reset) {
+    if (reset) {
+        metrics->pages_cache_write = atomic_exchange(&global_metrics.pages_cache_write, 0);
+        metrics->pages_cache_hit = atomic_exchange(&global_metrics.pages_cache_hit, 0);
+        metrics->pages_cache_miss = atomic_exchange(&global_metrics.pages_cache_miss, 0);
+        metrics->pages_cache_miss = atomic_exchange(&global_metrics.pages_cache_spill, 0);
+        metrics->read_txn_time_ns = atomic_exchange(&global_metrics.read_txn_time_ns, 0);
+        metrics->write_txn_time_ns = atomic_exchange(&global_metrics.write_txn_time_ns, 0);
+    } else {
+        metrics->pages_cache_write = atomic_load(&global_metrics.pages_cache_write);
+        metrics->pages_cache_hit = atomic_load(&global_metrics.pages_cache_hit);
+        metrics->pages_cache_miss = atomic_load(&global_metrics.pages_cache_miss);
+        metrics->pages_cache_miss = atomic_load(&global_metrics.pages_cache_spill);
+        metrics->read_txn_time_ns = atomic_load(&global_metrics.read_txn_time_ns);
+        metrics->write_txn_time_ns = atomic_load(&global_metrics.write_txn_time_ns);
+    }
+
+    return SQLITE_OK;
+}
+
+sqlite3_error sqlite3_collect_metrics(unsigned int event, void *pCtx, void *P, void *X) {
+    sqlite3_stmt* stmt;
+    sqlite3* connection;
+
+    int64_t stmt_time_ns;
+    int cache_hit, cache_miss, cache_write, cache_spill, _;
+    
+    switch (event) {
```

Comment:
Not sure if this is an existing convention, but would it be better to move these into the case where they're used?

Reply:
In C, you cannot declare variables inside a switch case. The only way would be to create a scope for that, but it is a bit weird? like

```C
switch (event) {
case SQLITE_TRACE_PROFILE:
    {
        ...
    }
    break;
 ...
 ```

Reply:
Ah of course, I forgot about this! If there's an existing convention, I think let's go with that, but I think if in future other interesting cases are added to this `switch`, it would be messy to have all of the variables declared above (especially if some names end up being shared!), so although somewhat ugly, creating a scope here would feel cleaner to me

Reply:
I brought back `if`s as they seem to be more readable.